src/base/emu-i386/simx86/cpu-emu.c

```diff
@@ -581,6 +581,24 @@ void Cpu2Reg (void)
 	REG(eflags),get_FLAGS(TheCPU.eflags),TheCPU.eflags);
 }
 
+void e_enter(void)
+{
+  if (CONFIG_CPUSIM)
+    fp87_load_fpstate(&vm86_fpu_state);
+  else {
+    // unmasked exception settings are emulated
+    TheCPU.fpuc = vm86_fpu_state.cwd;
+    vm86_fpu_state.cwd |= 0x3f;
+  }
+}
+
+void e_leave(void)
+{
+  if (CONFIG_CPUSIM)
+    fp87_save_fpstate(&vm86_fpu_state);
+  else
+    vm86_fpu_state.cwd = (vm86_fpu_state.cwd & ~0x3f) | (TheCPU.fpuc & 0x3f);
```

Comment:
Is this all it saves?
So jit doesn't really support fpu?

Reply:
JIT saves/restores from/to vm86_fpu_state elsewhere. Just need to sync with fpuc here.

But I think it would be cleaner in the end to eliminate the global vm86_fpu_state variable, and reintroduce TheCPU._fpstate for cpu-emu (removed in 7b31442).

Of course that would mean that kvm_enter etc. would need to pass a vm86_fpu_state pointer, but the state itself can be on the stack in `dpmi_set_pm()`.

Reply:
I added a patch to eliminate the global vm86_fpu_state.

The only odd one now is `kvm_fpu_update()`. It's identical to `kvm_enter()`  but I suppose `kvm_enter()` could be changed to also alter some KVM monitor fields that never change during vm86/dpmi.

In any case the port i/o function (which was buggy before depending on the backend) needs both a `get` and a `set`, so we'd need 4 functions per backend? enter/leave/update/getfpstate?

Reply:
so checking a bit further and checking Bochs (as commented), `fpu_reset()` does not need to read the FPU state as it completely resets the FPU (zeros all other values). The write to port 0xf0 doesn't need to clear the busy bit in the FPU status word, rather it acts like an EOI to the FPU after raising an exception (IRQ13); the code in bios.S does that in the int75 code. The busy bit has to do with fwait behaviour not exceptions.

So i'll get back to this tomorrow, also the failing test case in the CI for simulated vm86 + native DPMI

Reply:
I guess if you eliminate the global
state, then enter/leave are not needed,
so all you need is set_fpu_state()/
get_fpu_state()?

Also if you are at it, how about
(here or in another PR) to convert
that `if (cpuvm==` mess into the
cpu_backend ops? That of course
should not be limited to just fpu
ops, so quite a bit of a routine work
would be needed.

Reply:
using cpu_backend ops would be a great thing but indeed for another PR.
I've kept enter and leave as they'd allow some future optimizations/functionality (signalling a DPMI<->VM86 change to the backend).

Reply:
But its very strange to pass an
fpstate pointer to enter/leave.
Either we have some "messy"
func that syncs "some" state
or we have an explicit funcs like
get_fpu_state/set_fpu_state that
use the needed pointer.
For now it looks simplest to just
rename those along with adding
a parameter, and re-introduce
enter/leave when the need happens.

Also tests fail but devel is "clean"
wrt tests. So its either your patches,
or not rebased?

Reply:
The reason for the crash was a dirty fxstate, coming from fsave_to_fxsave from non-JIT cpuemu writing to the uninitialized fpu_state on the stack.

If you're happy with this, now with get/set functions I could clean up the git history a bit, if that would help (or just squash the regression fix?)

Reply:
actually something still bugs me a bit: in `cpu.c` you commented:
```
/* NOTE: this function does NOT memset the "unused" fxsave fields.
 * We preserve the previous fxsave context. */
void fsave_to_fxsave(const struct emu_fsave *fptr, struct emu_fpxstate *fxsave)
```

but then in `dnative.c`:

```
#else
    if ((scp_fpregs->status >> 16) == EMU_X86_FXSR_MAGIC)
      fpregs = &scp_fpregs->status + 1;
    else {
      static struct emu_fpxstate tmp;
      fsave_to_fxsave(fpregs, &tmp);
      fpregs = &tmp;
    }
#endif
    memcpy(fpstate, fpregs, sizeof(*fpstate));
```
so essentially the zeros of `tmp` get preserved, instead of letting `fsave_to_fxsave` write to the signal stack?

Note that we have essentially 2 scenarios, without and with SSE state, and the first and second require only copying the first 160 bytes resp. 288 bytes. And now that the intermediate state is temporary it could be reduced to 288 bytes plus a few auxiliary fields (e.g. denoting if it contains SSE state, or an explicit length (288 or 160), and it could also contain (if 160 bytes) the legacy fsave value of the tag field so you can skip conversion of that if it's not needed.

ie. a bit like the union I proposed earlier but instead it'll never be used directly by the asm instructions fnsave/frstor/fxsave/fxrstor?

Reply:
> so essentially the zeros of tmp get preserved, instead of letting fsave_to_fxsave write to the signal stack?

Thank!
I haven't tested that i386 code in
a hope you would. :) Which happened,
thanks again.
Not sure what do you mean by "write
to the signal stack" here, but I applied
the trivial fix. Does it help?

> ie. a bit like the union I proposed earlier but instead it'll never be used directly by the asm instructions
> fnsave/frstor/fxsave/fxrstor?

Do you mean, shorten memcpy() by
a dozen of bytes?

Reply:
The signal stack the sigcontext, it is what scp points to here, that is state that is filled in by the kernel when entering the handler.

My main goal isn't to shorten memcpy() necessarily. It's just that we don't have the SSE state in the simulator, and xmm8-xmm15 anywhere, so we shouldn't touch that state.

So what I would like ideally is this for the transition  from the simulator vm86 to native DPMI is that the x87 state (only!) is copied to the sigcontext, and the rest of that sigcontext isn't touched. Ie. the `get` from the simulator gets the x87 state, signals no SSE is there, and `set` then takes that into account.

For native DPMI -> simulator, the `get` function doesn't know that, so can copy 288 bytes just because the rest isn't used by any DOS code, but the similator `set` code just uses 160 bytes of those to fill in the x87 state.

The JIT has its private 512-byte fxstate, initialized by fxsave, and set only needs to update 288 bytes of those.
Similar for KVM, you can get the state from the VM, update it, then set.

Reply:
> So what I would like ideally is this for the transition from the simulator vm86 to native DPMI is that the x87
> state (only!) is copied to the sigcontext, and the rest of that sigcontext isn't touched. Ie. the get from the
> simulator gets the x87 state, signals no SSE is there, and set then takes that into account.

But its not the only way to achieve
that goal. My code can do it too. :)
And actually does, here's how:
from dnative you copy full fxsave
state to global state var.
From simx86 you don't copy full
fxsave, you call fsave_to_fxsave(),
but it explicitly preserves what was
previously copied from dnative.
So when you are back in dnative,
you actually CAN copy back the
full fxsave because its valid and
up-to-date.
So I believe the same goal is achieved,
but by a much more flexible means,
eg you have the global up-to-date
state too, which you can use for
save/restore (if save/restore would
ever be possible with dosemu2, but
thats another story :)

Reply:
by the way the quick fix is correct, I was confused about the signal stack in this scenario though, sorry! This place copies *from* the signal stack, not to it.

Reply:
> it explicitly preserves what was
> previously copied from dnative.

except the first time :)

The global fpstate was initialized by an explicit `fxsave` from the host (or fsave on old CPUs). This is what I seeked to eliminate as well, as the xmm8-xmm15 from there aren't relevant. Now we can of course `memset` this to 0 and fill in some fields in `fpu_reset()`, but that's the same global or local.

Then the first time we go from simx86 to DPMI, it's the host state plus that x87 state, and that gets copied to native DPMI. The magic values the kernel then put at the end of scp->fpregs are then overwritten.

Reply:
> Then the first time we go from simx86 to DPMI, it's the host state plus that x87 state, and that gets copied
> to native DPMI. The magic values the kernel then put at the end of scp->fpregs are then overwritten.

Our convert_from_fxsr() doesn't
seem to overwrite status.
Therefore, how about this:
```
--- a/src/dosext/dpmi/dnative/dnative.c
+++ b/src/dosext/dpmi/dnative/dnative.c
@@ -114,7 +114,8 @@ static void copy_to_dpmi(sigcontext_t *scp, cpuctx_t *s)
 #else
     /* i386: convert fxsave state to fsave state */
     convert_from_fxsr(scp->fpregs, &vm86_fpu_state);
-    if ((scp->fpregs->status >> 16) != EMU_X86_FXSR_MAGIC)
+    if (((scp->fpregs->status >> 16) != EMU_X86_FXSR_MAGIC) ||
+        (vm86_fpu_state.status >> 16) != EMU_X86_FXSR_MAGIC)
       return;
     fpregs = &scp->fpregs->status + 1;
 #endif
```

Reply:
EDITed above comment.

Reply:
the problem is the other magic values at the end. ie this part from
https://github.com/torvalds/linux/blob/master/arch/x86/include/uapi/asm/sigcontext.h

```
	union {
		__u32			padding2[12];
		struct _fpx_sw_bytes	sw_reserved;	/* Potential extended state is encoded here */
	};
```
kernel puts magic values there, they'll get overwritten by our zeros.
(ie. the zeros from the way global vm86_fpu_state is initialized by the C runtime, since `fxsave`, called at init time, doesn't touch that area)

Reply:
Yes, but what is the problem?
We save that to emu_fpstate, fully,
with all magic bytes. We should
restore them when returning to an
emulator.
What is the problem, if, for a time
being, we zero them out and go
to dpmi? That looks an advantage,
not a bug. :)
We restore when return to emu.

Reply:
I don't see anywhere we save/restore the emu FPU state. We used to but after realizing it was incomplete anyway, when returning to emu simply call `fesetenv(&dosemu_env);` (since 0e720ef096)

So essentially when entering DPMI and magic flags set to 0, the kernel will ignore the XSAVE state it put after fpregs and just use `fxrstor` on state we put there, ie the one we manually created with `fxsave` (or could manufacture in `fpu_reset` with a memset. So it does work.

I just would think it's cleaner to only update in the sigcontext what we need. e.g. for the CPU we update only 32-bit state but r8-r15 aren't changed. Yet for the FPU it's a more wholesale but still partial thing with xmm8-xmm15 but without ymm and zmm registers.

So just for consistency I think it's cleaner to only change the bits we need (x87 state + 32bit SSE state, but not xmm8-xmm15) in sigcontext->fpregs. That way the kernel will use xsave when entering the sighandler and when leaving, the magic values are always still there and it'll use xrstor on return.

Reply:
OK maybe you already removed it. :)
But being incomplete or not, it recovers
everything we could possibly corrupt,
as we also copy only 512bytes. 
So while fesetnev() doesn't hurt, we
can also have a private 512bytes
fxsave buffer in dnative.c.

> So just for consistency I think it's cleaner to only change the bits we need (x87 state + 32bit
> SSE state, but not xmm8-xmm15) in sigcontext->fpregs.

Lets specify what "change" is here.
If you want to not change the rest in
a dosemu-visible way, then the
aforementioned private buffer will
just do the trick perfectly: dosemu
will not see anything changed.
Of course if you insist on not changing
the rest AT ALL (even for dpmi) then
that's not what I expected, so my approach
currently does it differently.
But what does the AT ALL approach
give us over the "not change in any
dosemu-visible way"?

Reply:
Less copying, a smaller internal structure (`emu_fpstate` without xmm8-xmm15 and all the reserved bits), minimal messing around on the sigcontext and more consistency wrt CPU.

But I don't think the functionality changes, I just think it's cleaner without a global state, have the backends each maintain their own state, and copy from one backend to another when you change backend.
If backend 2 has SSE state but backend 1 has not, "2" can keep it, just copy the x87 state from 1 to 2.
And `fpu_reset` would need to clear the FPU state in both vm86 and DPMI backends if one is smaller than the other.

I mean for the CPU sigcontext we could also construct it manually: have
`static sigcontext_t dpmi_stack_frame;`
(so zero initialized)
and then `copy_to_dpmi` could do:
```
dpmi_stack_frame.es = get_es(s)
dpmi_stack_frame.ds = get_ds(s)
...
copy_context(scp, &dpmi_stack_frame)
```
but we don't, instead just updating the fields we need, and I like it that way :)

Anyway, we can agree to disagree here, it's not going to change anything from the user's point of view.

Reply:
I don't fully understand your proposal, too.
You want a static but and only partially
feed it to a sighandler. This won't overwrite
things, but next time you come from kvm
that DOES have fxsave.
How would you in that case propagate
that fxsave to sigcontext?

Reply:
I mean, how do you know the buffer
is just partially zeroed, or has full
fxsave?

Reply:
No, the static buffer as in the PR now is a workaround. I want to remove the static, and change `emu_fpstate` to only have a common first bit with emu_fpxstate (and thus remove `typedef struct emu_fpxstate emu_fpstate;` in `cpu.h`) We then have:

```
typedef struct {
  /* first 288 bytes match 32-bit FXSAVE format in 64bit mode  */
  uint16_t              cwd;
  uint16_t              swd;
  uint16_t              ftw;
  uint16_t              fop;
  uint32_t              fip;
  uint32_t              fcs;
  uint32_t              fdp;
  uint32_t              fds;
  uint32_t              mxcsr;
  uint32_t              mxcr_mask;
  struct emu_fpxreg     st[8];
  struct { uint32_t element[4]; } xmm[8];
  int from_fsave;
};
```
`fsave_to_fxsave` then sets `from_fsave` to `1`. If it's set, copy only cwd to fds and st[] (16 + 128 bytes) to scp->fpregs, if it's not set (from KVM), copy everything from cwd to xmm[] (288 bytes)to scp->fpregs.

I'll code something up.

Reply:
OK add flag, break the static_assert()
sizeof constraint...
No, please, don't code up that. :)
I really fail to see why do we need the
"incomplete" flag here.

Reply:
I think the `static_assert()` isn't that important but we probably differ here. With the above I view `emu_fpstate` as something internal, much like `cpuctx_t`, it doesn't need to match anything glibc or the kernel provide, just something we need to copy from and to.

Reply:
> it doesn't need to match anything glibc or the kernel provide, just something we need to copy from and to.

And replace a straight-forward
memcpy(.., .., sizeof(*fpstate));
with some weird offsetof() tricks?

Plus adding duality into state, 
and just for what? For running
dpmi in some "mixed" unknown
context instead of a clearly defined
or even synthesized context?
I really don't see the need.
Private buffer is good, that's what
you also add to simx86 - why not?

Reply:
However another alternative without a global vm86_fpu_state and still solving the cpu-emu simulator issue, is to change the cpu-emu code from this:

```
void e_set_fpu_state(const emu_fpstate *fpstate)
{
  if (CONFIG_CPUSIM)
    fp87_load_fpstate(fpstate);
  else {
    // unmasked exception settings are emulated
    TheCPU.fpuc = fpstate->cwd;
    TheCPU._fpstate = *fpstate;
    TheCPU._fpstate.cwd |= 0x3f;
  }
}

void e_get_fpu_state(emu_fpstate *fpstate)
{
  if (CONFIG_CPUSIM)
    fp87_save_fpstate(fpstate);
  else {
    *fpstate = TheCPU._fpstate;
    fpstate->cwd = (fpstate->cwd & ~0x3f) | (TheCPU.fpuc & 0x3f);
  }
}
```
to this
```
void e_set_fpu_state(const emu_fpstate *fpstate)
{
  TheCPU._fpstate = *fpstate;
  if (CONFIG_CPUSIM)
    fp87_load_fpstate(fpstate);
  else {
    // unmasked exception settings are emulated
    TheCPU.fpuc = fpstate->cwd;
    TheCPU._fpstate.cwd |= 0x3f;
  }
}

void e_get_fpu_state(emu_fpstate *fpstate)
{
  *fpstate = TheCPU._fpstate;
  if (CONFIG_CPUSIM)
    fp87_save_fpstate(fpstate);
  else
    fpstate->cwd = (fpstate->cwd & ~0x3f) | (TheCPU.fpuc & 0x3f);
}
```
so the whole state is copied, even if the simulator doesn't use the SSE bits (of course the simulator could use `TheCPU._fpstate` instead of its own `static long double _fparea[8];` but that's another thing).

Then `dnative.c` could still only copy the first 288 bytes but that's completely local.

Reply:
How will that solve the "not-zeroed
initial state" problem that broke tests
here, or do you mean some other
problem?
In general, if sim uses fsave format,
then loading fxsave and then saving
fsave mixed with that, doesn't look
right to me.

Simple question: is there any real
problem still available here if we
remove the requirement of running
dpmi in a "mixed" state? Because
to me it looks this is the only source
of all problems. Everything else is
solved, or am I missing something?

Reply:
It would essentially eliminate the mixed state with that change, in all backends, and the "non-zero state" would vanish, on `set` you make the simulator save the SSE state, and on `get` you get exactly the same state back, only x87 stuff has changed.

Sim doesn't really use fsave format, it was just convenient to code it that way because of the tag word conversion, it can save directly to fxsave also.

So no more mixed state in DPMI, just only copying the first 288 bytes (always, well except for old i386 of course) would allow the kernel to use its magic fields.

Reply:
Unifying all back-ends to fxsave
format is definitely good too.
But of course you are still going
to overwrite the real SSE state in
a sighandler with "dummy" SSE
state from SIM, right? That's fine
with me, though IIRC this is what
you were fighting against. :)

Also can there be a pathological
cases like vm86() on non-fxsr CPU
and sim-dpmi with fxsr? Then we'll
probably get to the same problem.

Anyway, please unify simx86 to fxsr,
that's definitely an option.

Reply:
So I added two patches, one, 99071c3, to preserve SSE state in all backends and a second one 8c33f6d to decrease  the size of emu_fpstate (which you probably don't want -- static_assert needs offsetof -- but I was still curious if it passed all tests).

I believe I got all pathological cases though, the weirdest is for native DPMI without fxsr, since the SSE state can't be put in the sigcontext then, it has to be stored in a separate file-scope static variable.

Reply:
So you want to keep the fsave
state shared and fxsave state
private this time?

Reply:
Not sure what exactly your patches
do now, but having only fsave shared
may be a deal. As I really don't expect
someone in DOS wants to propagate
SSE state from PM to RM.

Reply:
no, everything, both fxsave and fsave is private, 99071c3 made sure the SSE state is preserved in addition to the x87 state, which was already preserved.

Reply:
So if I understand correctly, now
you save the fxsave state privately
and load later either fully, or converting
to fsave first?

> no, everything, both fxsave and fsave is private

Yes but what gets propagated between
back-ends? (I call that a shared part)

Reply:
yes both fxsave and fsave (if needed, e.g. in dpmi_native it's i386 only) are private. The fields in there (both fxsave and fsave) get filled in by the `set` and taken out by `get`. Then it gets loaded later (by an explicit fxrstor/frstor instruction, or by copying to sigcontext).

In KVM there are no explicit static private structures though, it's taken out and put into the VM straight away, but for the others we cannot do that.

shared part = first 288 bytes of fxsave structure. Just convenient for memcpy, but it just has everything we need.

So the full fxsave only features when we actually use the FXSAVE instruction (in true_vm86 and JIT).

Reply:
But how do you know which
one of those private structs is
valid?
I propose to propagate the flag
from getter to setter - similar to
what you did before, except to
not add such a flag to any struct.
Passing it between 2 funcs is fine.
That way you will eventually have
the full fxsave in a private buffer,
so no need for 2 private buffers.

Reply:
The fxsave one is always valid. The fsave one is used on demand, and functions like a cache, to avoid converting back and forth for e.g. every single vm86 call.

But indeed using a flag in the setters/getters is another possibility and may eliminate the need to convert from fsave to fxsave for sure (and fxsave to fsave only for i386 in native dpmi since the kernel needs that).

Reply:
(I mean valid at get and set time)

Reply:
ok so playing with this further I think it's simplest to use a union with fsave and fxsave but with the flag being `config.cpufxsr`. And on x86_64 fsave is never used (it could be a union with one field there).

It's theoretically possible to have an i386 kernel that does not save fxsave on a CPU that can, but that's simple enough, force config.cpufxsr = 0 once you see the kernel doesn't save it in sigcontext.

To keep things consistent, the simulator can then save/restore its state directly into fsave or fxsave depending on config.cpufxsr.

I was playing with that but did not have time to finish that today.

Reply:
> To keep things consistent, the simulator can then save/restore its state directly into fsave or fxsave
> depending on config.cpufxsr.

But that's the problem.
Simulator shouldn't care, and what
if we eventually have tcg as a simulator?
I really think its better to use the separate
flag in getters/setters and try to "accumulate"
the full fxsave state eventually.
I.e. keep exchanging the fxsave state as
now, but have an "incomplete" flag to say
that only the fsave part is trusted.
This will solve the "first-time" problem, and
on non-first time you'll accumulate the full
fxsave state, so that flag will clear.
What do you think about that?
That's a very minimal change against the
current code, as it already does the same
thing.